### PR TITLE
Remove Mapper.updateFieldType()

### PR DESCRIPTION
**Pull request by romseygeek**
When we had multiple mapping types, an update to a field in one type had to be
propagated to the same field in all other types.  This was done using the 
`Mapper.updateFieldType()` method, called at the end of a merge.  However, now
that we only have a single type per index, this method is unnecessary and can
be removed.

Relates to #41059

### CODE REVIEW

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java [hunk 1]
```diff
@@ -85,7 +85,8 @@
     public static ParentJoinFieldMapper getMapper(MapperService service) {
         MetaJoinFieldMapper.MetaJoinFieldType fieldType =
             (MetaJoinFieldMapper.MetaJoinFieldType) service.fieldType(MetaJoinFieldMapper.NAME);
-        return fieldType == null ? null : fieldType.getMapper();
+        return fieldType == null ? null :
+            (ParentJoinFieldMapper) service.fieldMapper(fieldType.getJoinField());
```
Comment (romseygeek) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java:89]: ParentJoinFieldMapper was using updateFieldType() to update its linked MetaJoinFieldMapper.  This requires a small bit of hackery to get round, which I hope to fix in a followup by moving the various filter factory methods from the field mapper to the field type.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 2]
```diff
@@ -590,6 +584,10 @@ public MappedFieldType fieldType(String fullName) {
         return fieldTypes.get(fullName);
     }
 
+    public Mapper fieldMapper(String fieldName) {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:587]: This is unfortunately necessary as a shim to get round the ParentJoinFieldMapper issue above.  Hopefully removed asap.

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:587]: Could we avoid adding this new method and call this inline in `ParentJoinFieldMapper`? All the methods like `documentMapper()` are public, so it seems do-able.

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: ++ have reworked the mocking (which sent me down a surprisingly deep rabbit hole, but I think it's fixed now!)

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 3]
```diff
@@ -439,7 +431,9 @@ private boolean assertMappersShareSameFieldType() {
             Collections.addAll(fieldMappers, mapper.mapping().metadataMappers);
             MapperUtils.collect(mapper.root(), new ArrayList<>(), fieldMappers, new ArrayList<>());
             for (FieldMapper fieldMapper : fieldMappers) {
-                assert fieldMapper.fieldType() == fieldTypes.get(fieldMapper.name()) : fieldMapper.name();
+                assert Objects.equals(fieldMapper.fieldType(), fieldTypes.get(fieldMapper.name())) :
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: ++

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 4]
```diff
@@ -812,10 +812,6 @@ private static void parseDynamicValue(final ParseContext context, ObjectMapper p
             builder = createBuilderFromDynamicValue(context, token, currentFieldName);
         }
         Mapper mapper = builder.build(builderContext);
-        if (existingFieldType != null) {
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:815]: ++

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 5]
```diff
@@ -295,10 +293,6 @@ public boolean shouldCache(Query query) {
         MapperService mapperService = mapperServiceMock();
         when(mapperService.getIndexSettings()).thenReturn(indexSettings);
         when(mapperService.hasNested()).thenReturn(false);
-        DocumentMapper mapper = mock(DocumentMapper.class);
-        when(mapper.typeText()).thenReturn(new Text(TYPE_NAME));
-        when(mapper.type()).thenReturn(TYPE_NAME);
-        when(mapperService.documentMapper()).thenReturn(mapper);
```
Comment (romseygeek) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:301]: I wasted several hours trying to work out why my mocked DocumentMapper class wasn't returning the DocumentFieldMapper instance I was asking it to.  This appears to be unnecessary, so I've removed it

**Review by jtibshirani — APPROVED** [review]
Looks good to me!

**Comment by romseygeek** [comment]
The backport on this is a bit complicated due to default types still being present (but unused!) in 7x, so I will open a separate PR.

**Pull request merged**